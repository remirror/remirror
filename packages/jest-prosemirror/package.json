{
  "name": "jest-prosemirror",
  "version": "3.0.0-beta.4",
  "description": "Write expressive tests for your prosemirror editor",
  "keywords": [
    "jest",
    "prosemirror",
    "snapshots",
    "tests"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-prosemirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-prosemirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-prosemirror.js",
      "require": "./dist/jest-prosemirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-prosemirror-environment.js",
      "require": "./environment/dist/jest-prosemirror-environment.cjs"
    },
    "./package.json": "./package.json",
    "./serializer": {
      "types": "./dist-types/serializer.d.ts",
      "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
    }
  },
  "main": "./dist/jest-prosemirror.cjs",
  "module": "./dist/jest-prosemirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment",
    "serializer"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core-constants": "3.0.0-beta.3",
    "@remirror/core-helpers": "4.0.0-beta.3",
    "@remirror/core-types": "3.0.0-beta.4",
    "@remirror/core-utils": "3.0.0-beta.4",
    "@remirror/pm": "3.0.0-beta.4",
    "@testing-library/dom": "^9.3.0",
    "@types/stringify-object": "^3.3.1",
    "chalk": "^4.1.0",
    "jest-matcher-utils": "^29.5.0",
    "prosemirror-schema-basic": "^1.2.3",
    "prosemirror-tables": "^1.3.7",
    "prosemirror-test-builder": "^1.1.1",
    "stringify-object": "^3.3.0",
    "test-keyboard": "2.0.7-beta.4"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-prosemirror.d.ts",
        "import": "./dist/jest-prosemirror.js",
        "require": "./dist/jest-prosemirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-prosemirror-environment.d.ts",
        "import": "./environment/dist/jest-prosemirror-environment.js",
        "require": "./environment/dist/jest-prosemirror-environment.cjs"
      },
      "./package.json": "./package.json",
      "./serializer": {
        "types": "./serializer/dist/jest-prosemirror-serializer.d.ts",
        "require": "./serializer/dist/jest-prosemirror-serializer.cjs"
      }
    },
    "main": "./dist/jest-prosemirror.cjs",
    "module": "./dist/jest-prosemirror.js",
    "type": "module",
    "types": "./dist/jest-prosemirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts",
      "serializer.cts"
    ]
  }
}
