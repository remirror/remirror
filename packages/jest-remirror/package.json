{
  "name": "jest-remirror",
  "version": "3.0.0-beta.6",
  "description": "A text editor for react built with prosemirror",
  "keywords": [
    "prosemirror",
    "react",
    "text editor"
  ],
  "homepage": "https://github.com/remirror/remirror/tree/HEAD/packages/jest-remirror",
  "repository": {
    "type": "git",
    "url": "https://github.com/remirror/remirror.git",
    "directory": "packages/jest-remirror"
  },
  "license": "MIT",
  "contributors": [
    "Ifiok Jr. <ifiokotung@gmail.com>"
  ],
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "import": "./dist/jest-remirror.js",
      "require": "./dist/jest-remirror.cjs"
    },
    "./environment": {
      "types": "./dist-types/environment.d.ts",
      "import": "./environment/dist/jest-remirror-environment.js",
      "require": "./environment/dist/jest-remirror-environment.cjs"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/jest-remirror.cjs",
  "module": "./dist/jest-remirror.js",
  "types": "./dist-types/index.d.ts",
  "files": [
    "dist",
    "dist-types",
    "environment"
  ],
  "scripts": {
    "build": "remirror-cli build"
  },
  "dependencies": {
    "@babel/runtime": "^7.22.3",
    "@remirror/core": "3.0.0-beta.6",
    "@remirror/dom": "3.0.0-beta.6",
    "@remirror/pm": "3.0.0-beta.4",
    "@remirror/preset-core": "3.0.0-beta.6",
    "@testing-library/dom": "^9.3.0",
    "@types/sanitize-html": "^2.9.0",
    "jest-prosemirror": "3.0.0-beta.4",
    "sanitize-html": "^2.10.0",
    "tiny-warning": "^1.0.3"
  },
  "devDependencies": {
    "@remirror/cli": "1.1.0-beta.0",
    "@types/node": "^16.18.34",
    "jest": "^29.5.0"
  },
  "peerDependencies": {
    "@types/node": "*",
    "jest": ">= 24.9.0"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    },
    "jest": {
      "optional": true
    }
  },
  "publishConfig": {
    "access": "public",
    "exports": {
      ".": {
        "types": "./dist/jest-remirror.d.ts",
        "import": "./dist/jest-remirror.js",
        "require": "./dist/jest-remirror.cjs"
      },
      "./environment": {
        "types": "./environment/dist/jest-remirror-environment.d.ts",
        "import": "./environment/dist/jest-remirror-environment.js",
        "require": "./environment/dist/jest-remirror-environment.cjs"
      },
      "./package.json": "./package.json"
    },
    "main": "./dist/jest-remirror.cjs",
    "module": "./dist/jest-remirror.js",
    "type": "module",
    "types": "./dist/jest-remirror.d.ts"
  },
  "@remirror": {
    "tsconfigs": {
      "src": {
        "compilerOptions": {
          "types": [
            "node",
            "jest"
          ]
        }
      }
    }
  },
  "preconstruct": {
    "entrypoints": [
      "index.ts",
      "environment.ts"
    ]
  }
}
