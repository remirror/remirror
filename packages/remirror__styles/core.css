/**
 * Styles extracted from: packages/remirror__theme/src/core-theme.ts
 */
.remirror-editor.ProseMirror {
  word-wrap: break-word;
  white-space: pre-wrap;
  white-space: break-spaces;
  position: relative;
  font-variant-ligatures: none;
  font-feature-settings: 'liga' 0;
  overflow-y: scroll;
}

.remirror-editor.ProseMirror pre {
  white-space: pre-wrap;
}

.remirror-editor.ProseMirror li {
  position: relative;
}

.remirror-editor.ProseMirror[contenteditable='false'] {
  white-space: normal;
}

.remirror-editor.ProseMirror[contenteditable='true'] {
  white-space: pre-wrap;
}

.remirror-editor.ProseMirror hr {
  border-color: #2e2e2e;
}

/* Protect against generic img rules. See also https://github.com/ProseMirror/prosemirror-view/commit/aaa50d592074c8063fc2ef77907ab6d0373822fb */

.remirror-editor.ProseMirror img.ProseMirror-separator {
  display: inline !important;
  border: none !important;
  margin: 0 !important;
}
.remirror-editor.ProseMirror-hideselection *::-moz-selection {
  background: transparent;
  color: inherit;
}
.remirror-editor.ProseMirror-hideselection *::selection {
  background: transparent;
  color: inherit;
}
.remirror-editor.ProseMirror-hideselection *::-moz-selection {
  background: transparent;
  color: inherit;
}
.remirror-editor.ProseMirror-hideselection {
  caret-color: transparent;
}
.remirror-editor .ProseMirror-selectednode {
  outline: 2px solid #8cf;
}
/* Make sure li selections wrap around markers */
.remirror-editor li.ProseMirror-selectednode {
  outline: none;
}
.remirror-editor li.ProseMirror-selectednode:after {
  content: '';
  position: absolute;
  left: -32px;
  right: -2px;
  top: -2px;
  bottom: -2px;
  border: 2px solid #8cf;
  pointer-events: none;
}
