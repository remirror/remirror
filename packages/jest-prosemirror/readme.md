# jest-prosemirror

[![npm](https://img.shields.io/npm/dm/jest-prosemirror.svg?&logo=npm)](https://www.npmjs.com/package/jest-prosemirror)

## The problem

You want to write tests for some of your prosemirror editor but you don't know where to start. You know you should avoid testing implementation details and just want to be sure that your commands and plugins produce the correct underlying prosemirror state.

## This solution

`jest-prosemirror` takes inspiration from the [`testing-library`](https://github.com/testing-library/react-testing-library) mantra and enables you to write more intuitive tests for your prosemirror editor.

## Installation

```bash
yarn add jest-prosemirror # yarn
pnpm add jest-prosemirror # pnpm
npm install jest-prosemirror # npm
```

## Getting started

### Quick setup

For a quick setup add the following to your jest.config.js file.

```js
module.exports = {
  setupFilesAfterEnv: ['jest-prosemirror/environment'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

This will automatically:

- Add the jest assertions `toTransformNode` and `toEqualProsemirrorNode`.

If you are using typescript then add this to your `tsconfig.json` file for global type support.

```json
{
  "compilerOptions": {
    "types": ["jest-prosemirror"]
  }
}
```

### Manual setup

Create a `jest.framework.dom.ts` file and add the following

```ts
// jest.framework.dom.ts

import { prosemirrorMatchers } from 'jest-prosemirror';

// Add jest-prosemirror assertions
expect.extend(prosemirrorMatchers);
```

In your `jest.config.js` file add this to the configuration

```js
module.exports = {
  setupFilesAfterEnv: ['<rootDir>/jest.framework.dom.ts'],
  testEnvironment: 'jsdom', // Required for dom manipulation
};
```

## Snapshot serializer

This package exports a serializer for better snapshot testing of prosemirror primitives. To set this up add the following to your `jest.config.js` file.

```js
module.exports = {
  snapshotSerializers: ['jest-prosemirror/serializer'],
};
```

Alternatively, you can add the following to your `jest.framework.dom.ts` file.

```ts
// jest.framework.dom.ts

import { prosemirrorSerializer } from 'jest-prosemirror';

// Add the serializer for use throughout all the configured test files.
expect.addSnapshotSerializer(prosemirrorSerializer);
```

## Jest Matchers

### `toTransformNode`

A utility from jest-prosemirror which tests that a command transforms the prosemirror node in the desired way.

```ts
import { doc, p, schema, strong } from 'jest-prosemirror';
import { removeMark } from 'remirror/core/utils';

test('remove the mark', () => {
  const type = schema.marks.bold;
  const from = doc(p(strong('<start>bold<end>')));
  const to = doc(p('bold'));

  expect(removeMark({ type })).toTransformNode({ from, to });
});
```

This tests that mark has been removed by the provided command.

The `to` property is optional and can be left blank to test that the node is identical after the transform.

### `toEqualProsemirrorNode`

Tests that two prosemirror documents are equal. Pass in the expected document and it checks that they are the same.

```ts
import { createEditor, doc, p } from 'jest-prosemirror';
import { removeNodeAtPosition } from 'remirror/core/utils';

test('remove block top level node at specified position', () => {
  const {
    state: { tr },
  } = createEditor(doc(p('x'), p('one')));
  const newTr = removeNodeAtPosition({ pos: 3, tr });

  expect(newTr).not.toBe(tr);
  expect(newTr.doc).toEqualProsemirrorNode(doc(p('x')));
});
```

## `createEditor`

Create a test prosemirror editor.

The call to create editor can be chained with various commands to enable testing of the editor at each step without the need for intermediate holding variables.

```ts
import { createEditor, doc, p } from 'jest-remirror';
import { suggest } from 'prosemirror-suggest';

test('`keyBindings`', () => {
  const keyBindings = {
    Enter: jest.fn((params: SuggestKeyBindingParameter) => {
      params.command();
    }),
  };

  const plugin = suggest({
    char: '@',
    name: 'at',
    keyBindings,
    matchOffset: 0,
    createCommand:
      ({ view }) =>
      () =>
        view.dispatch(view.state.tr.insertText('awesome')),
  });

  createEditor(doc(p('<cursor>')), { plugins: [plugin] })
    .insertText('@')
    .press('Enter')
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorNode(doc(p('@awesome')));
    });
});
```

### Parameters

#### taggedDoc - `TaggedProsemirrorNode`

The tagged prosemirror node to inject into the editor.

#### options - `CreateEditorOptions`

The following options which include all [DirectEditorProps](http://prosemirror.net/docs/ref/#view.DirectEditorProps) except for `state`.

| **Property** | **Type** | **Default** | **Description** |
| --- | :-: | :-: | --- |
| `autoClean` | `boolean` | `true` | Whether to auto remove the editor from the dom after each test. It is advisable to leave this unchanged. |
| `plugins` | `Plugin[]` | `[]` | The plugins that the test editor should use. |
| `rules` | `InputRule[]` | `[]` | The input rules that the test editor should use. |

### Returns

#### start - `number`

The start of the current selection.

#### end - `number | undefined`

The end of the current selection.

#### selection - `Selection`

The current prosemirror selection

#### schema - `EditorSchema`

The prosemirror schema.

#### state - `EditorState`

The prosemirror state.

#### view - `EditorView`

The prosemirror view.

#### overwrite - `(doc: TaggedProsemirrorNode) => ReturnType<typeof createEditor>`

Overwrite all the current content within the editor.

**`doc`** - the new content to use.

#### command - `(command: CommandFunction) => ReturnType<typeof createEditor>`

Run the command within the prosemirror editor.

```ts
test('commands are run', () => {
  createEditor(doc(p('<cursor>')))
    .command((state, dispatch) => {
      if (dispatch) {
        dispatch(state.tr.insertText('hello'));
      }
    })
    .callback((content) => {
      expect(content.state.doc).toEqualProsemirrorDocument(doc(p('hello')));
    });
});
```

**`command`** - the command function to run.

#### insertText - `(text: string) => ReturnType<typeof createEditor>`

Insert text into the editor at the current position.

**`text`** - the text to insert

#### jumpTo - `(start: 'start' | 'end' | number, end?: number) => ReturnType<typeof createEditor>`

Jump to the specified position in the editor.

**`start`** - a number position or the shorthand 'start' | 'end'. **`[end]`** - the option end position of the new selection.

#### shortcut - `(mod: string) => ReturnType<typeof createEditor>`

Type a keyboard shortcut - e.g. `Mod-Enter`.

**NOTE** This only simulates the events. For example an `Mod-Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`mod`** - the keyboard shortcut to type

#### press - `(char: string) => ReturnType<typeof createEditor>`

Simulate a keypress which is run through the editor's key handlers.

**NOTE** This only simulates the events. For example an `Enter` would run all enter key handlers but not actually create a new line. I'd welcome a pull request to fix this shortcoming.

**`char`** - the character to type

#### fire - `(params: Omit<FireEventAtPositionParameter, 'view'>) => ReturnType<typeof createEditor>`

Fire an event in the editor. This api is not the most well tested and can be quite flakey.

**`params`** - the fire event parameters

#### callback - `(fn: (content: ReturnValueCallbackParameter) => void) => ReturnType<typeof createEditor>`

Callback function which receives the `start`, `end`, `state`, `view`, `schema` and `selection` properties and allows for easier testing of the current state of the editor.

## Acknowledgements

This package borrows very heavily from [prosemirror-test-builder](https://www.npmjs.com/package/prosemirror-test-builder)
