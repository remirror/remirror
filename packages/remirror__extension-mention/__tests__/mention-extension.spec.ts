import { jest } from '@jest/globals';
import { pmBuild } from 'jest-prosemirror';
import { extensionValidityTest, renderEditor } from 'jest-remirror';
import {
  createCoreManager,
  ItalicExtension,
  MentionChangeHandler,
  MentionExtension,
  MentionOptions,
} from 'remirror/extensions';
import { hideConsoleError } from 'testing';
import { htmlToProsemirrorNode, prosemirrorNodeToHtml } from '@remirror/core';

extensionValidityTest(MentionExtension, { matchers: [] });

describe('schema', () => {
  const { schema } = createCoreManager([
    new MentionExtension({ matchers: [{ char: '@', name: 'at' }] }),
  ]);
  const attributes = { id: 'test', label: '@test', name: 'testing' };

  const { mention, p, doc } = pmBuild(schema, {
    mention: { markType: 'mention', ...attributes },
  });

  it('creates the correct dom node', () => {
    expect(prosemirrorNodeToHtml(p(mention(attributes.label)))).toMatchInlineSnapshot(`
      <p>
        <a
          class="mention mention-testing"
          data-mention-id="test"
          data-mention-name="testing"
        >
          @test
        </a>
      </p>
    `);
  });

  it('parses the dom structure and finds itself', () => {
    const node = htmlToProsemirrorNode({
      schema,
      content: `<a class="mention mention-at" data-mention-id="${attributes.id}" data-mention-name="${attributes.name}">${attributes.label}</a>`,
    });
    const expected = doc(p(mention(attributes.label)));

    expect(node).toEqualProsemirrorNode(expected);
  });

  describe('extraAttributes', () => {
    const custom = 'test';
    const { schema } = createCoreManager([
      new MentionExtension({
        matchers: [],
        extraAttributes: { 'data-custom': { default: null } },
      }),
    ]);

    const { doc, p, mention } = pmBuild(schema, {
      mention: { markType: 'mention', ['data-custom']: custom, ...attributes },
    });

    it('parses the dom structure and finds itself with custom attributes', () => {
      const node = htmlToProsemirrorNode({
        schema,
        content: `<a class="mention mention-at" data-custom="${custom}" data-mention-id="${attributes.id}" data-mention-name="${attributes.name}">${attributes.label}</a>`,
      });

      const expected = doc(p(mention(attributes.label)));

      expect(node).toEqualProsemirrorNode(expected);
    });
  });
});

/**
 * Create the mention extension with an optional `onChange` handler.
 */
function create(options: MentionOptions) {
  const extension = new MentionExtension({ ...options });
  const editor = renderEditor([extension, new ItalicExtension()]);
  const { add, view, manager, commands } = editor;
  const { doc, p } = editor.nodes;
  const { mention } = editor.attributeMarks;

  return { add, doc, p, mention, view, manager, commands, editor, extension };
}

describe('`createSuggesters`', () => {
  const options: MentionOptions = {
    matchers: [
      { char: '#', name: 'tag', mentionClassName: 'custom' },
      { char: '@', name: 'at', mentionClassName: 'custom' },
      { char: '+', name: 'plus', mentionClassName: 'custom' },
    ],
  };

  const id = 'mention';
  const label = `@${id}`;

  it('uses default noop callbacks and does nothing', () => {
    const noop: any = jest.fn();
    const { add, doc, p, extension } = create(options);
    extension.addHandler('onChange', noop);

    add(doc(p('<cursor>')))
      .insertText(`This ${label} `)
      .callback(({ state }) => {
        expect(state).toContainRemirrorDocument(p(`This ${label} `));
        expect(noop).toHaveBeenCalledTimes(8);
      });
  });

  const change: any = jest.fn();

  // The default `onChange` handler. Make sure to include this in future work.
  const onChange: MentionChangeHandler = jest.fn<MentionChangeHandler>((props, command) => {
    if (props.exitReason) {
      command();
    }

    if (props.changeReason) {
      change();
    }
  });

  const { add, doc, p, mention, view, editor, extension } = create(options);
  extension.addHandler('onChange', onChange);
  const mentionMark = mention({ id, label, name: 'at' });

  it('should support exits', () => {
    add(doc(p('<cursor>'))).insertText(`${label} `);
    expect(view.state).toContainRemirrorDocument(p(mentionMark(label), ' '));
  });

  it('should handle joined text separated by space', () => {
    add(doc(p('hello <cursor>friend'))).insertText(`${label} `);
    expect(view.state).toContainRemirrorDocument(p('hello ', mentionMark(label), ' friend'));
  });

  it('can split mentions', () => {
    const splitMention = mention({ id: '123', label: '@123', name: 'at' });

    add(doc(p(splitMention('@1<cursor>23'))));

    editor.insertText(' ');
    expect(view.state).toContainRemirrorDocument(
      p(mention({ id: '1', label: '@1', name: 'at' })('@1'), ' 23'),
    );
  });

  it('removes invalid mentions', () => {
    const splitMention = mention({ id: '123', label: '@123', name: 'at' });
    add(doc(p(splitMention('@<cursor>123')))).insertText(' ');

    expect(view.state).toContainRemirrorDocument(p('@ 123'));
  });

  it('decorates split mentions', () => {
    add(doc(p('hello <cursor>friend'))).insertText(`${label}`);

    expect(view.dom.innerHTML).toMatchInlineSnapshot(`
      <p>
        hello
        <a class="suggest suggest-at">
          @mentionfriend
        </a>
      </p>
    `);
  });

  it('injects the mention at the correct place', () => {
    add(doc(p('<cursor>'))).insertText(`This ${label} `);

    expect(view.state).toContainRemirrorDocument(p('This ', mentionMark(label), ' '));
    expect(change).toHaveBeenCalledTimes(id.length);
  });

  it('supports deleting content', () => {
    add(doc(p('<cursor>')))
      .insertText(`@1`)
      .dispatchCommand(({ state, dispatch }) => {
        if (dispatch) {
          dispatch(state.tr.delete(1, state.selection.head));
        }

        return true;
      })
      .callback(({ doc }) => {
        expect(doc).toMatchSnapshot();
      });
  });

  it('supports multiple characters', () => {
    const labelFn = (char: string) => `${char}${id}`;

    const hashMark = mention({ id, label: labelFn('#'), name: 'tag' });
    const plusMark = mention({ id, label: labelFn('+'), name: 'plus' });
    const atMark = mention({ id, label: labelFn('@'), name: 'at' });

    add(doc(p('<cursor>'))).insertText(`This ${labelFn('#')} `);
    expect(view.state).toContainRemirrorDocument(p('This ', hashMark(labelFn('#')), ' '));

    add(doc(p('<cursor>'))).insertText(`This ${labelFn('+')} `);
    expect(view.state).toContainRemirrorDocument(p('This ', plusMark(labelFn('+')), ' '));

    add(doc(p('<cursor>'))).insertText(`This ${labelFn('@')} `);
    expect(view.state).toContainRemirrorDocument(p('This ', atMark(labelFn('@')), ' '));
  });
});

describe('pasteRules', () => {
  const options = {
    matchers: [
      { char: '#', name: 'tag' },
      { char: '@', name: 'at' },
      { char: '+', name: 'plus' },
    ],
  };

  const { add, doc, p, mention } = create(options);

  it('supports pasted content', () => {
    add(doc(p('<cursor>')))
      .paste(p('This is text @hello'))
      .callback(({ state }) => {
        const expectedMention = mention({ id: '@hello', name: 'at', label: '@hello' })('@hello');
        const expected = p('This is text ', expectedMention);

        expect(state).toContainRemirrorDocument(expected);
      });
  });

  it('supports pasting multiple matchers', () => {
    const at = mention({ id: '@hello', name: 'at', label: '@hello' })('@hello');
    const plus = mention({ id: '+awesome', name: 'plus', label: '+awesome' })('+awesome');
    const tag = mention({ id: '#12', name: 'tag', label: '#12' })('#12');

    add(doc(p('<cursor>')))
      .paste(p('#12 +awesome @hello'))
      .callback((content) => {
        expect(content.state).toContainRemirrorDocument(p(tag, ' ', plus, ' ', at));
      });
  });
});

describe('commands', () => {
  const options = {
    matchers: [
      { char: '#', name: 'tag' },
      { char: '@', name: 'at' },
      { char: '+', name: 'plus' },
    ],
  };

  const { add, doc, p, mention, view, commands } = create(options);
  const attributes = { id: 'test', label: '@test', name: 'at', appendText: '' };

  describe('createMention', () => {
    hideConsoleError(true);

    it('replaces text at the current position by default', () => {
      add(doc(p('This is ', '<cursor>')));
      commands.createMention(attributes);

      expect(view.state).toContainRemirrorDocument(
        p('This is ', mention(attributes)(attributes.label)),
      );
    });

    it('replaces text at the specified position', () => {
      add(doc(p('This is ', '<cursor>')));
      commands.createMention({ ...attributes, range: { from: 1, to: 1, cursor: 1 } });

      expect(view.state).toContainRemirrorDocument(
        p(mention(attributes)(attributes.label), 'This is '),
      );
    });

    it('throws when invalid config passed into the command', () => {
      add(doc(p('This is ', '<cursor>')));

      // @ts-expect-error
      expect(() => commands.createMention()).toThrowErrorMatchingSnapshot();

      // @ts-expect-error
      expect(() => commands.createMention({})).toThrowErrorMatchingSnapshot();

      expect(() =>
        commands.createMention({ ...attributes, id: '' }),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        commands.createMention({ ...attributes, label: '' }),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        commands.createMention({ ...attributes, name: 'invalid' }),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        // @ts-expect-error
        commands.createMention({ ...attributes, name: undefined }),
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('interactions with input rules', () => {
  const options: MentionOptions = {
    matchers: [{ char: '@', name: 'at', mentionClassName: 'custom' }],
  };

  const { add, doc, p, mention, view } = create(options);

  it('should skip input rules when mention active', () => {
    const mentionMark = mention({ id: 'mention', label: '@mention', name: 'at' });
    const { state } = add(doc(p('123 ', mentionMark('@men_tion<cursor>')))).insertText('_ ');

    expect(state.doc).toEqualRemirrorDocument(doc(p('123 ', mentionMark('@men_tion'), '_ ')));
  });

  it('should skip input rules when mention suggestion active', () => {
    add(doc(p('123 '))).insertText('@a_bc_ ');

    expect(view.dom.innerHTML).toMatchInlineSnapshot(`
      <p>
        123
        <a class="suggest suggest-at">
          @a_bc_
        </a>
      </p>
    `);
  });

  it('should skip input for overlapping sections', () => {
    add(doc(p('_123 '))).insertText('@abc_ ');

    expect(view.dom.innerHTML).toMatchInlineSnapshot(`
      <p>
        _123
        <a class="suggest suggest-at">
          @abc_
        </a>
      </p>
    `);
  });

  it('should support longer matches', () => {
    add(doc(p('123 '))).insertText('@a_bc_ this should be preserved');

    expect(view.dom.innerHTML).toMatchInlineSnapshot(`
      <p>
        123 @a_bc_ this should be preserved
      </p>
    `);
  });
});

describe('forwardDeletes', () => {
  const options: MentionOptions = {
    matchers: [{ char: '@', name: 'at', mentionClassName: 'custom' }],
  };

  const { add, doc, p, mention, view } = create(options);

  it('should support deleting forward', () => {
    const mentionMark = mention({ id: 'mention', label: '@mention', name: 'at' });
    add(doc(p('abc <cursor>', mentionMark('@mention')))).forwardDelete();

    expect(view.state.doc).toEqualRemirrorDocument(doc(p('abc mention')));
  });
});

describe('onClick', () => {
  const options = {
    matchers: [
      { char: '#', name: 'tag' },
      { char: '@', name: 'at' },
      { char: '+', name: 'plus' },
    ],
  };

  const { add, doc, p, mention, view, extension } = create(options);

  it('responds to clicks', () => {
    const clickHandler: any = jest.fn(() => true);
    extension.addHandler('onClick', clickHandler);
    const atMention = mention({ id: '@hello', name: 'at', label: '@hello' })('@hello');
    const node = p('first ', atMention);
    add(doc(node));

    view.someProp('handleClickOn', (fn) => fn(view, 10, node, 1, {} as MouseEvent, false));
    expect(clickHandler).toHaveBeenCalledTimes(1);
    expect(clickHandler).toHaveBeenCalledWith(
      {},
      expect.objectContaining({
        from: 7,
        to: 13,
        text: '@hello',
        mark: expect.any(Object),
      }),
    );

    view.someProp('handleClick', (fn) => fn(view, 3, {} as MouseEvent));
    expect(clickHandler).toHaveBeenCalledTimes(1);
  });
});
