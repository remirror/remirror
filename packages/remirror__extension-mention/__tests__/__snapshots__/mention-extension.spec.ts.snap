// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`createSuggesters\` supports deleting content 1`] = `
Prosemirror node: {
  "type": "doc",
  "content": [
    {
      "type": "paragraph"
    }
  ]
}
`;

exports[`commands createMention throws when invalid config passed into the command 1`] = `
"An error occurred within an extension. More details should be made available.

Invalid configuration attributes passed to the MentionExtension command.

For more information visit https://remirror.io/docs/errors#rmr0100"
`;

exports[`commands createMention throws when invalid config passed into the command 2`] = `
"An error occurred within an extension. More details should be made available.

Invalid configuration attributes passed to the MentionExtension command.

For more information visit https://remirror.io/docs/errors#rmr0100"
`;

exports[`commands createMention throws when invalid config passed into the command 3`] = `
"An error occurred within an extension. More details should be made available.

Invalid configuration attributes passed to the MentionExtension command.

For more information visit https://remirror.io/docs/errors#rmr0100"
`;

exports[`commands createMention throws when invalid config passed into the command 4`] = `
"An error occurred within an extension. More details should be made available.

Invalid configuration attributes passed to the MentionExtension command.

For more information visit https://remirror.io/docs/errors#rmr0100"
`;

exports[`commands createMention throws when invalid config passed into the command 5`] = `
"An error occurred within an extension. More details should be made available.

The name 'invalid' specified for this command is invalid. Please choose from: ["tag","at","plus"].

For more information visit https://remirror.io/docs/errors#rmr0100"
`;

exports[`commands createMention throws when invalid config passed into the command 6`] = `
"An error occurred within an extension. More details should be made available.

Invalid configuration attributes passed to the MentionExtension command.

For more information visit https://remirror.io/docs/errors#rmr0100"
`;
